File: agent/python/perper/model/agent.py
import asyncio
from .async_locals import get_cache_service, get_notification_service, get_local_agent, get_instance, get_execution
from perper.protocol.standard import PerperAgent

startup_function_name = 'Startup'

def get_agent():
    return PerperAgent(get_local_agent(), get_instance())

def call_function(delegate, parameters):
    return agent_call_function(get_agent(), delegate, parameters)

def call_action(delegate, parameters):
    return agent_call_action(get_agent(), delegate, parameters)

async def start_agent(agent, parameters):
    instance = get_cache_service().generate_name(agent)
    get_cache_service().instance_create(instance, agent)
    model = PerperAgent(agent, instance)
    result = await agent_call_function(model, startup_function_name, parameters)
    return (model, result)

async def agent_call_function(agent, delegate, parameters):
    result = await __agent_call(agent, delegate, parameters, get_cache_service().call_read_result)

    if result is None:
        return None
    elif len(result) == 1:
        return result[0]
    else:
        return tuple(result)

async def agent_call_action(agent, delegate, parameters):
    await __agent_call(agent, delegate, parameters, get_cache_service().call_check_result)

async def __agent_call(agent, delegate, parameters, result_handler):
    call = get_cache_service().generate_name(delegate)

    call_notification_task = asyncio.create_task(get_notification_service().get_call_result_notification(call))  # HACK: Workaround bug in fabric
    get_cache_service().call_create(call, agent.agent, agent.instance, delegate, get_local_agent(), get_instance(), parameters)

    (k, n) = await call_notification_task
    result = result_handler(call)
    get_cache_service().call_remove(call)
    get_notification_service().consume_notification(k)

    return result

def agent_destroy(agent):
    get_cache_service().instance_destroy(agent.instance)

File: agent/dotnet/src/Perper/Model/Agent.cs
using System;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;

using Perper.Protocol.Cache.Standard;
using Perper.Protocol.Extensions;

namespace Perper.Model
{
    public class Agent : IAgent
    {
        public Agent(PerperAgent rawAgent) => RawAgent = rawAgent;

        public PerperAgent RawAgent { get; }

        public async Task<TResult> CallFunctionAsync<TResult>(string functionName, object[] parameters)
        {
            var results = await CallAsync(functionName, parameters).ConfigureAwait(false);

            if (typeof(ITuple).IsAssignableFrom(typeof(TResult)))
            {
                return (TResult)Activator.CreateInstance(typeof(TResult), results);
            }
            else if (typeof(object[]) == typeof(TResult))
            {
                return (TResult)(object)results;
            }
            else if (results.Length >= 1)
            {
                return (TResult)results[0];
            }
            else
            {
                return default!;
            }
        }

        public async Task CallActionAsync(string actionName, object[] parameters)
        {
            await CallAsync(actionName, parameters).ConfigureAwait(false);
        }

        private async Task<object[]> CallAsync(string @delegate, object[] parameters)
        {
            var call = AsyncLocals.CacheService.GenerateName(@delegate);

            var callNotificationTask = AsyncLocals.NotificationService.GetCallResultNotification(call).ConfigureAwait(false); // HACK: Workaround bug in fabric
            await AsyncLocals.CacheService.CallCreate(call, RawAgent.Agent, RawAgent.Instance, @delegate, AsyncLocals.Agent, AsyncLocals.Instance, parameters).ConfigureAwait(false);
            var (notificationKey, _) = await callNotificationTask;

            var results = await AsyncLocals.CacheService.CallReadResult(call).ConfigureAwait(false);
            await AsyncLocals.NotificationService.ConsumeNotification(notificationKey).ConfigureAwait(false); // TODO: Consume notifications and save state entries in a smarter way
            await AsyncLocals.CacheService.CallRemove(call).ConfigureAwait(false);

            return results;
        }

        public async Task Destroy()
        {
            await AsyncLocals.CacheService.InstanceDestroy(RawAgent.Instance).ConfigureAwait(false);
        }
    }
}